InformationScripting/src/sources/AstSource.cpp

```diff
@@ -0,0 +1,83 @@
+/***********************************************************************************************************************
+**
+** Copyright (c) 2011, 2015 ETH Zurich
+** All rights reserved.
+**
+** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "AstSource.h"
+
+#include "OOModel/src/declarations/Class.h"
+
+#include "ModelBase/src/nodes/Node.h"
+
+#include "../graph/Graph.h"
+#include "../graph/InformationNode.h"
+
+namespace InformationScripting {
+
+AstSource& AstSource::instance()
+{
+	static AstSource instance;
+	return instance;
+}
+
+void AstSource::init()
+{
+	auto astHash = [](const InformationNode* n) -> QPair<std::size_t, bool> {
+		if (n->contains("ast"))
+		{
+			Model::Node* nodePtr = (*n)["ast"];
+			return {std::hash<Model::Node*>()(nodePtr), true};
+		}
+		return {0, false};
+	};
+
+	Graph::registerNodeHash(astHash);
+}
+
+AstQuery* AstSource::createClassesQuery(Model::Node* target, QStringList args)
+{
+	return new AstQuery(AstQuery::QueryType::Classes, target, args);
```

Comment:
Hmm, I'm beginning to doubt the need for a factory at all. Why not create the queries directly?